ssd1289/ssd1289.go

```
@@ -5,44 +5,35 @@ package ssd1289
 
 import (
 	"image/color"
-	"machine"
 	"time"
+
+	"tinygo.org/x/drivers/internal/legacy"
+	"tinygo.org/x/drivers/internal/pin"
 )
 
 type Bus interface {
 	Set(data uint16)
 }
 
 type Device struct {
-	rs  machine.Pin
-	wr  machine.Pin
-	cs  machine.Pin
-	rst machine.Pin
+	rs  pin.OutputStruct
+	wr  pin.OutputStruct
+	cs  pin.OutputStruct
+	rst pin.OutputStruct
 	bus Bus
 }
 
 const width = int16(240)
 const height = int16(320)
 
-func New(rs machine.Pin, wr machine.Pin, cs machine.Pin, rst machine.Pin, bus Bus) Device {
-	d := Device{
-		rs:  rs,
-		wr:  wr,
-		cs:  cs,
-		rst: rst,
+func New(rs, wr, cs, rst pin.Output, bus Bus) *Device {
```

Big compile-time bang here, as we return pointer now.

---

I guess returning a pointer would agree with the ways of other drivers. Are we OK with breaking our users on this? We'd be opening the door to breaking other drivers that do this. It's fine by me if we all agree

---

On second thought, I'd perform this change in one fell swoop on all drivers

---

We did this multiple times, so 🤷 As I understood practice is "if it breaks in compile time, then fine". We don't want to break ppl silently, keeping API, but changing behaviour.

---

Note: the function signature change is not a breaking change unless someone thought of storing the function in a data structure, which I believe is very, VERY unlikely. Most (read as all) people use drivers as seen in examples so this should not be a breaking change

---

That's true, good point, compile-time breaks only if they specified type, not inferred it.

So, if they infer type -- it just works for them, a variable stops being struct and becomes pointer to struct.
If they define variable type explicitly, it will break for them in compile time. I'd say that's acceptable.

I may try and dig older PRs there we did such change w/o much fuzz, so I understood we are fine with this.

---

We should probably have always returned a pointer type, but like several things subsequent decisions make for some inconsistencies.

I think that returning pointer now is fine, as it does break compile (no silent failures) and also is an easy fix.

---

Alright, we did it before, no point to stop doing it now :) Pointer it is.